kafka_consumer/datadog_checks/kafka_consumer/kafka_consumer.py

```diff
@@ -74,81 +26,97 @@ def __init__(self, name, init_config, instances):
+                            assert isinstance(partition, int)
+
+    def _init_check_based_on_kafka_version(self):
+        """Determine whether to use old legacy KafkaClient implementation or the new KafkaAdminClient implementation.
 
-        # Fetch Kafka consumer offsets
-        try:
-            self._get_consumer_offsets()
-        except Exception:
-            self.log.exception("There was a problem collecting consumer offsets from Kafka.")
-            # don't raise because we might get valid broker offsets
+        The legacy version of this check uses the KafkaClient and handrolls things like looking up the GroupCoordinator,
+        crafting the offset requests, handling errors, etc.
 
-        # Fetch the broker highwater offsets
-        try:
-            if len(self._consumer_offsets) < self._context_limit:
-                self._get_highwater_offsets()
-            else:
-                self.warning("Context limit reached. Skipping highwater offset collection.")
-        except Exception:
-            self.log.exception("There was a problem collecting the highwater mark offsets.")
-            # Unlike consumer offsets, fail immediately because we can't calculate consumer lag w/o highwater_offsets
-            raise
+        The new implementation uses the KafkaAdminClient which lets us offload most of the Kafka-specific bits onto the
+        kafka-python library, which is used by many other tools and reduces our maintenance burden.
+
+        Unfortunately, the KafkaAdminClient requires brokers >= 0.10.0, so we split the check into legacy and new code.
 
-        total_contexts = len(self._consumer_offsets) + len(self._highwater_offsets)
-        if total_contexts >= self._context_limit:
-            self.warning(
-                """Discovered %s metric contexts - this exceeds the maximum number of %s contexts permitted by the
-                check. Please narrow your target by specifying in your kafka_consumer.yaml the consumer groups, topics
-                and partitions you wish to monitor.""",
-                total_contexts,
-                self._context_limit,
-            )
+        Furthermore, we took the opportunity to simplify the new code by dropping support for:
+        1) Zookeeper-based offsets. These have been deprecated since Kafka 0.9.
+        2) Kafka brokers < 0.10.2. It is impossible to support monitor_unlisted_consumer_groups on these older brokers
+        because they do not provide a way to determine the mapping of consumer groups to topics. For details, see
+        KIP-88.
 
-        # Report the metrics
-        self._report_highwater_offsets(self._context_limit)
-        self._report_consumer_offsets_and_lag(self._context_limit - len(self._highwater_offsets))
+        To clarify: This check still allows fetching offsets from zookeeper/older kafka brokers, it just uses the
+        legacy code path."""
+        kafka_version = self.instance.get('kafka_client_api_version')
```

Comment:
if this is going to run on check initializations it should check if `self.sub_check` is already set and do nothing if that's the case

Reply:
This method only succeeds once so it's ok I think, right? Once `self.sub_check` is set the methods returns and is never ran again